fml/platform/darwin/weak_nsobject.h

```diff
@@ -0,0 +1,287 @@
+
+#include <stdlib.h>
+#include "flutter/fml/compiler_specific.h"
+#include "flutter/fml/logging.h"
+#include "flutter/fml/memory/ref_counted.h"
+#include "flutter/fml/memory/ref_ptr.h"
+#include "flutter/fml/memory/thread_checker.h"
+
+namespace debug {
+struct DebugThreadChecker {
+  FML_DECLARE_THREAD_CHECKER(checker);
+};
+}  // namespace debug
+
+// WeakNSObject<> is patterned after scoped_nsobject<>, but instead of
+// maintaining ownership of an NSObject subclass object, it will nil itself out
+// when the object is deallocated.
+//
+// WeakNSProtocol<> has the same behavior as WeakNSObject, but can be used
+// with protocols.
+//
+// Example usage (fml::WeakNSObject<T>):
+//   WeakNSObjectFactory factory([[Foo alloc] init]);
+//   WeakNSObject<Foo> weak_foo;  // No pointer
+//   weak_foo = factory.GetWeakNSObject() // Now a weak reference is kept.
+//   [weak_foo description];  // Returns [foo description].
+//   foo.reset();  // The reference is released.
+//   [weak_foo description];  // Returns nil, as weak_foo is pointing to nil.
+//
+//
+// Implementation wise a WeakNSObject keeps a reference to a refcounted
+// WeakContainer. There is one unique instance of a WeakContainer per watched
+// NSObject, this relationship is maintained via the ObjectiveC associated
+// object API, indirectly via an ObjectiveC CRBWeakNSProtocolSentinel class.
+//
+// Threading restrictions:
+// - Several WeakNSObject pointing to the same underlying object must all be
+//   created and dereferenced on the same thread;
+// - thread safety is enforced by the implementation, except:
+//     - it is allowed to destroy a WeakNSObject on any thread;
+// - the implementation assumes that the tracked object will be released on the
+//   same thread that the WeakNSObject is created on.
+//
+// fml specifics:
+// WeakNSObjects can only originate from a |WeakNSObjectFactory| (see below), though WeakNSObjects
+// are copyable and movable.
+//
+// WeakNSObjects are not in general thread-safe. They may only be *used* on
+// a single thread, namely the same thread as the "originating"
+// |WeakNSObjectFactory| (which can invalidate the WeakNSObjects that it
+// generates).
+namespace fml {
+
+// Forward declaration, so |WeakNSObject<NST>| can friend it.
+template <typename NST>
+class WeakNSObjectFactory;
+
+// WeakContainer keeps a weak pointer to an object and clears it when it
+// receives nullify() from the object's sentinel.
+class WeakContainer : public fml::RefCountedThreadSafe<WeakContainer> {
+ public:
+  explicit WeakContainer(id object, std::shared_ptr<debug::DebugThreadChecker> checker = nullptr)
+      : checker_(checker), object_(object) {}
+
+  id object() {
+    CheckThreadSafety();
+    return object_;
+  }
+
+  void nullify() {
+    CheckThreadSafety();
+    object_ = nil;
+  }
+
+ private:
+  friend fml::RefCountedThreadSafe<WeakContainer>;
+  ~WeakContainer() {}
+
+  std::shared_ptr<debug::DebugThreadChecker> checker_;
+  id object_;
```

Comment:
This doesn't look ARC-safe to me; this is a strong reference under ARC unless I'm misremembering the semantics of `id` in ARC. I was confused by this so I looked at the history, and although:

> This PR mostly copies the weak_nsobject from chromium, at the same hash that we copied the ARC/MRC compatible scoped_nsobject: https://chromium.googlesource.com/chromium/src/+/fd625125b8a6a3aceaf09993a5f74cfe5368b17f

that doesn't appear to have been an ARC-safe version. See https://codereview.chromium.org/2943263003 which landed much later.

Reply:
Thanks for catching this. I have adopted the change and updated the NSWeakObjectFactory accordingly.
I have also added a new unittests to test the NSWeakObject under ARC.
Also migrated a XCTest file that contains NSWeakObject to ARC.